csharp/src/Client/AdbcDataReader.cs

```diff
@@ -0,0 +1,324 @@
+        }
+
+        public override object GetValue(int ordinal)
+        {
+            return GetValue(this.recordBatch?.Column(ordinal), ordinal);
+        }
+
+        public override int GetValues(object[] values)
+        {
+            int count = Math.Min(FieldCount, values.Length);
+            for (int i = 0; i < count; i++)
+            {
+                values[i] = GetValue(i);
+            }
+            return count;
+        }
+
+        public override bool IsDBNull(int ordinal)
+        {
+            return GetValue(ordinal) == null;
+        }
+
+        public override bool NextResult()
+        {
+            this.recordBatch = ReadNextRecordBatchAsync().Result;
+
+            if (this.recordBatch != null)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public new void Dispose()
+        {
+            this.recordBatch?.Dispose();
+        }
+
+        public override bool Read()
+        {
+            if (this.recordBatch != null && this.currentRowInRecordBatch < this.recordBatch.Length - 1)
+            {
+                this.currentRowInRecordBatch++;
+                return true;
+            }
+
+            this.recordBatch = ReadNextRecordBatchAsync().Result;
+
+            return this.recordBatch != null;
+        }
+
+        public override DataTable GetSchemaTable()
+        {
+            if (this.schema != null)
+            {
+                return SchemaConverter.ConvertArrowSchema(this.schema, this.adbcCommand.AdbcStatement);
+            }
+            else
+            {
+                throw new InvalidOperationException("Cannot obtain schema table because the Arrow schema is null");
+            }
+        }
+
+#if NET5_0_OR_GREATER
+        public override Task<DataTable> GetSchemaTableAsync(CancellationToken cancellationToken = default)
+        {
+            return Task.Run(() => GetSchemaTable(), cancellationToken);
+        }
+#endif
+
+        public ReadOnlyCollection<DbColumn> GetColumnSchema()
+        {
+            if (this.schema != null)
+            {
+                List<DbColumn> dbColumns = new List<DbColumn>();
+
+                foreach (Field f in this.schema.FieldsList)
+                {
+                    Type t = SchemaConverter.ConvertArrowType(f);
+
+                    if(
+                        f.DataType.TypeId == Types.ArrowTypeId.Double ||
+                        f.DataType.TypeId == Types.ArrowTypeId.Float ||
+                        f.DataType.TypeId == Types.ArrowTypeId.Decimal128 ||
+                        f.DataType.TypeId == Types.ArrowTypeId.Decimal256
+                    )
+                    {
+                        int precision = Convert.ToInt32(f.Metadata["precision"]);
+                        int scale = Convert.ToInt32(f.Metadata["scale"]);
+                        dbColumns.Add(new AdbcColumn(f.Name, t, f.IsNullable, precision, scale));

```

Comment:
Btw, can we store the database physical type in dbColumns?  In ado.net the providerType indicate the physical type of database.  But after the arrow-adbc convert, the providerType become arrowTypes related.

Reply:
DbColumn doesn't have a ProviderType property like a DataColumn in a DataTable does.  That said, knowing that information could be helpful so I added the `ArrowType` property to AdbcColumn.